File: greylist.pl
#!/usr/bin/perl

# This plugin implements greylisting using a SQLite database
# See http://www.sqlite.org for more information on SQLite

use strict;
use warnings;

use Carp;
use DBI;
use DBD::SQLite;
use Domain::PublicSuffix;
use Time::Local;
use NetAddr::IP::Lite qw(:lower);

# Seconds until greylisting is deactivated for 'triplet'
my $mindelay = 170;
# Seconds a 'triplet' is valid (no more greylisting)
# Raised from 28800 to 604800 - 20100620 - stevem
my $maxvalid = 604800;
# Default netmask for sender
my $mask = "255.255.255.0";
# Greylisting reply
my $output = "E451 Greylisting active. Your mail will be accepted on the next attempt. This is NOT an error.\n";

# By default, greylisting is active
my $greylist = 1;

# Handle IPv6
my $tcpremoteip;
if ($ENV{'PROTO'} eq 'TCP')  { $tcpremoteip = $ENV{'TCPREMOTEIP'} }
if ($ENV{'PROTO'} eq 'TCP6') { $tcpremoteip = "0.0.0.0" }

# Make sure TCPREMOTEHOST is set
my $tcpremotehost = "";
if (defined($ENV{'TCPREMOTEHOST'})) {
  $tcpremotehost = $ENV{'TCPREMOTEHOST'};
}
if (defined($ENV{'TCP6REMOTEHOST'})) {
  $tcpremotehost = $ENV{'TCP6REMOTEHOST'};
}

# Net::Netmask object
my $ipobj = NetAddr::IP::Lite->new($tcpremoteip, $mask);
my $network = $ipobj->network();
   $network =~ s/\/\d+$//ix;

# Domain::PublicSuffix object
my $suffix = Domain::PublicSuffix->new;
my $rdns;
if (defined($suffix->get_root_domain($tcpremotehost))) {
  $rdns = "".$suffix->get_root_domain($tcpremotehost);
} else {
  $rdns = "";
}

# Relayclients will not be greylisted, of course.
if ( (defined($ENV{'RELAYCLIENT'})) ||
     (defined($ENV{'TRUSTCLIENT'})) ||
     (not(-w "/var/qmail/control/greylist/")) ) {
  print "\n";
  exit(0);
}

# Allow temporary disable
if (-f "/var/qmail/control/greylist/disable") {
  print "\n";
  exit(0);
}

my $i;
my $now = timelocal(localtime);
my $sdomain = "";
if ($ENV{'SMTPMAILFROM'} =~ /\@(.*)$/x) {
  $sdomain = "$1";
}

my $dbh = DBI->connect("dbi:SQLite:dbname=/var/qmail/control/greylist/sqlite.db","","")
          || croak "Cannot connect: $DBI::errstr";

my $res = $dbh->selectall_arrayref( "SELECT timestamp FROM main WHERE ipaddr = \"$network\" AND mail = \"$sdomain\" AND rcpt = \"$ENV{'SMTPRCPTTO'}\"" );

foreach (@$res) {
  foreach my $i (0..$#$_) {
    # Check if the 'triplet' is older then $mindelay and younger then $maxvalid
    if ( ($_->[$i] >= ($now - $maxvalid)) && ($_->[$i] <= ($now - $mindelay)) ) {
      $greylist = 0;
      $output = "\n";
    }
  }
}

# Record the current 'triplet' to allow next attempt
$dbh->do( "INSERT INTO main VALUES (\"$network\",\"$sdomain\",\"$ENV{'SMTPRCPTTO'}\",\"$now\",\"$rdns\")" );

# Delete old entries in the DB (older than $maxvalid) and clean out the DB
if (($$ % 100) == 0) {
  print STDERR getppid()." Vacuuming greylist database\n";
  $dbh->do( "DELETE FROM main WHERE timestamp < $now - $maxvalid" );
  $dbh->do( "VACUUM" );
}

if ($greylist) {
  print STDERR getppid()." Added to greylist\n";
  print $output;
} else {
  print STDERR getppid()." Passed greylist test\n";
  print $output;
}

$dbh->disconnect;
exit;

File: greylist.go
package main

import "database/sql"
import "fmt"
import _ "github.com/mattn/go-sqlite3"
import "github.com/c-robinson/iplib"
import "golang.org/x/net/publicsuffix"
import "golang.org/x/sys/unix"
import "log"
import "net"
import "os"
import "strings"
import "strconv"
import "time"

const configdir = "/var/qmail/control/greylist"
const message = "E451 Greylisting active. Your mail will be accepted on the next attempt."

// int64 because epoch is also int64
const mindelay int64 = 170
const maxvalid int64 = 604800

func main() {
	if env_defined("RELAYCLIENT") ||
		env_defined("TRUSTCLIENT") ||
		file_exists(configdir+"/disable") ||
		!file_exists(configdir+"/sqlite.db") ||
		(unix.Access(configdir, unix.W_OK) != nil) {
		fmt.Println()
		os.Exit(0)
	}

	var remoteip string
	var remotenet string
	var remotehost string
	var remotedomain string

	// Put client IP into `remoteip`
	// Depending on IPv4/IPv6/TLS, it can be in different places
	if env_defined("TCPREMOTEIP") {
		remoteip = os.Getenv("TCPREMOTEIP")
	}
	if env_defined("TCP6REMOTEIP") {
		remoteip = os.Getenv("TCP6REMOTEIP")
	}
	if env_defined("SSLREMOTEIP") {
		remoteip = os.Getenv("SSLREMOTEIP")
	}

	// Extract the subnet from `remoteip` and put it into `remotenet`
	ipobj := net.ParseIP(remoteip)
	if ipobj.To4() != nil {
		// remoteip is IPv4
		remotenet = fmt.Sprint(iplib.NewNet(ipobj, 24).NetworkAddress()) + "/24"
	} else {
		// remoteip is IPv6
		remotenet = fmt.Sprint(iplib.NewNet(ipobj, 56).NetworkAddress()) + "/56"
	}

	// Same for `remotehost`
	if env_defined("TCPREMOTEHOST") {
		remotehost = os.Getenv("TCPREMOTEHOST")
	}
	if env_defined("TCP6REMOTEHOST") {
		remotehost = os.Getenv("TCP6REMOTEHOST")
	}
	if env_defined("SSLREMOTEHOST") {
		remotehost = os.Getenv("SSLREMOTEHOST")
	}

	// If `remotehost` is populated, find the ETLD+1
	if len(remotehost) > 0 {
		remotedomain, _ = publicsuffix.EffectiveTLDPlusOne(remotehost)
	}

	// Extract the sender domain
	var senderdomain string
	if env_defined("SMTPMAILFROM") {
		addrparts := strings.Split(os.Getenv("SMTPMAILFROM"), "@")
		if len(addrparts) == 2 {
			senderdomain = addrparts[1]
		}
	}

	// Read the recipient
	var recipient string
	if env_defined("SMTPRCPTTO") {
		recipient = os.Getenv("SMTPRCPTTO")
	}

	// Check the current time
	var epoch = (time.Now()).Unix()

	// Open database
	db, err := sql.Open("sqlite3", configdir+"/sqlite.db")
	if err != nil {
		fmt.Println(message)
		log.Fatal(err)
		err = db.Ping()
		if err != nil {
			fmt.Println(message)
			log.Fatal(err)
		}
	}
	defer db.Close()

	// Clean up cruft
	if (epoch % 60) == 0 {
		fmt.Fprintf(os.Stderr, "%d Vacuuming greylist database\n", os.Getppid())
		_, err = db.Exec(`DELETE FROM main WHERE timestamp < ` + strconv.FormatInt((epoch-maxvalid), 10) + `; VACUUM`)
		if err != nil {
			fmt.Fprintf(os.Stderr, "%d Failed to vacuum database: %s\n", os.Getppid(), err)
		}
	}

	stmt, err := db.Prepare("SELECT COUNT(*) FROM main WHERE" +
		"(timestamp >= ? AND timestamp <= ?)" +
		"AND mail = ? AND rcpt = ? AND" +
		"(ipaddr = ? OR rdns = ?)")
	if err != nil {
		fmt.Fprintf(os.Stderr, "%d Failed to prepare query: %s\n", os.Getppid(), err)
	}
	var count int
	//	fmt.Printf("SELECT COUNT(*) FROM main WHERE (timestamp >= %d AND timestamp <= %d) AND mail = \"%s\" AND rcpt = \"%s\" AND (ipaddr = \"%s\" OR rdns = \"%s\");\n", epoch - maxvalid, epoch + mindelay, senderdomain, recipient, remotenet, remotedomain)
	err = stmt.QueryRow(epoch-maxvalid, epoch-mindelay, senderdomain, recipient, remotenet, remotedomain).Scan(&count)
	if err != nil {
		fmt.Println(message)
		log.Fatal(err)
	}

	// Found an entry in the greylist DB
	if count > 0 {
		fmt.Fprintf(os.Stderr, "%d IP %s passed greylist test\n", os.Getppid(), remoteip)
		fmt.Println()
		os.Exit(0)
	}

	// Add client to the database
	_, err = db.Exec(`INSERT INTO main VALUES ("` + remotenet + `","` + senderdomain + `","` + recipient + `","` + strconv.FormatInt(epoch, 10) + `","` + remotedomain + `")`)
	if err != nil {
		fmt.Fprintf(os.Stderr, "%d Failed to insert into greylist DB: %s\n", os.Getppid(), err)
	} else {
		fmt.Fprintf(os.Stderr, "%d IP %s added to greylist (%s -> %s)\n", os.Getppid(), remoteip, senderdomain, recipient)
	}
	fmt.Println(message)

	os.Exit(0)
}

func env_defined(key string) bool {
	value, exists := os.LookupEnv(key)
	_ = value

	return exists
}

func file_exists(filename string) bool {
	_, err := os.Stat(filename)
	return !os.IsNotExist(err)
}
